File: Problem89_GrayCode.py
from typing import List

class Solution:
    def grayCode(self, n: int) -> List[int]:
        binCode = [] 
        subSet = []
        def backtrack(pos):
            if pos>=n:
                binCode.append(subSet.copy())
                return
            for c in ['0','1']:
                subSet.append(c)
                backtrack(pos+1)
                subSet.pop()
        backtrack(0)
        for i in range(len(binCode)):
            for j in range(n-1,0,-1):
                if binCode[i][j-1] == '1':
                    binCode[i][j] = '1' if binCode[i][j]=='0' else '0'
        num = ["".join(listNum) for listNum in binCode]
        res = [int(s, 2) for s in num]
        return res
    
if __name__ == "__main__":
    solve = Solution()
    print(solve.grayCode(2))

File: Problem89_GrayCode.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<string> combinations;
    string oneComb;
    int bin2Dec(string s) {
        int res=0, n=s.length();
        for (int i=0;i<n;i++) {
            res += (s[i]-'0')*pow(2,n-i-1);
        }
        return res;
    }
    void backtrack(int pos, int n) {
        if (pos>=n) {
            combinations.push_back(oneComb);
            return;
        }
        for (char i='0';i<='1';i++) {
            oneComb.push_back(i);
            backtrack(pos+1, n);
            oneComb.pop_back();
        }
    };
    vector<int> grayCode(int n) {
        backtrack(0, n);
        for (string s : combinations) cout << s << " ";
        cout << endl;
        for (int j=0;j<combinations.size();j++) {
            for (int i=n-1;i>=1;i--) {
                if (combinations[j][i-1]=='1') {
                    combinations[j][i] = (combinations[j][i]=='0') ? '1' : '0';
                }
            }
        }
        vector<int> res;
        for (string s : combinations) {
            res.push_back(bin2Dec(s));
        }
        return res;
    }
};

int main() {
    const int n=4;
    Solution solve;
    vector<int> res = solve.grayCode(n);
    for (int x : res) cout << x << " ";
}